### PR TITLE
[SYCL] Re-apply task sequence changes

### PR DESCRIPTION
**Pull request by aejjehint**
This change re-apply the task sequence changes from PR #12453 that were reverted in #14359 now that the spec has been approved and the back-end implementation is ready. The spec is also moved from proposed to experimental.

### CODE REVIEW

**Review by smanna12 — APPROVED** [review]

File: clang/test/CodeGenSYCL/intel-task-sequence.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+// Test that SPIR-V codegen generates the expected LLVM struct name for the
+// TaskSequenceINTEL type
+
+namespace __spv {
+  struct __spirv_TaskSequenceINTEL;
+} // namespace __spv
+
+// CHECK: @_Z4func{{.*}}(target("spirv.TaskSequenceINTEL")
+void func(__spv::__spirv_TaskSequenceINTEL *task_seq) {}
```
Comment (mdtoguchi) [inline clang/test/CodeGenSYCL/intel-task-sequence.cpp:11]: Looks to be some EOF marker problem here.

Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-task-sequence.cpp:11]: oops. I missed this? Thanks @mdtoguchi for catching!

**Review by mdtoguchi — APPROVED** [review]
Looks OK for Driver

**Comment by smanna12** [comment]
@steffenlarsen, @AlexeySachkov, @GarveyJoe, could you please review this PR? Thank you

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_properties.hpp [hunk 2]
```diff
@@ -12,11 +12,15 @@
 #include <sycl/ext/oneapi/properties/property.hpp>
 #include <sycl/ext/oneapi/properties/property_value.hpp>
 
+#include <cstdint>     // for uint16_t
+#include <type_traits> // for true_type
```
Comment (steffenlarsen) [inline sycl/include/sycl/ext/intel/experimental/fpga_kernel_properties.hpp:16]: Nit; We don't add the comments anymore, so you can remove them.

File: sycl/test-e2e/TaskSequence/mult-and-add.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  myQueue.submit([&](handler &cgh) {
+    auto res_acc = res_buf.get_access<access::mode::write>(cgh);
+    cgh.single_task(
+        [=](kernel_handler kh) { res_acc[0] = mult_and_add(1, 2, 3, 4); });
+  });
+  myQueue.wait();
+
+  assert(result == (1 * 2 + 3 * 4));
+}
```
Comment (steffenlarsen) [inline sycl/test-e2e/TaskSequence/mult-and-add.cpp:41]:
```suggestion
}

```

File: sycl/test-e2e/TaskSequence/producer-consumer.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  });
+  myQueue.wait();
+
+  // Check result:
+  int sum = 0;
+  for (int i = 0; i < kSize; ++i)
+    sum += i * i + 42;
+  assert(result == sum);
+  return 0;
+}
```
Comment (steffenlarsen) [inline sycl/test-e2e/TaskSequence/producer-consumer.cpp:61]:
```suggestion
}

```

File: sycl/test/check_device_code/task_sequence_intel_balanced.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      }
+
+      for (int i = 0; i < kSize; i++) {
+        results_acc[i] = arrayAddTask.get();
+      }
+    });
+  });
+  myQueue.wait();
+  return 0;
+}
```
Comment (steffenlarsen) [inline sycl/test/check_device_code/task_sequence_intel_balanced.cpp:46]:
```suggestion
}

```

File: sycl/test/check_device_code/task_sequence_intel_explicit_get.cpp [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      }
+
+      for (int i = 0; i < kSize; i++) {
+        results_acc[i] = arrayAddTask.get();
+      }
+    });
+  });
+  myQueue.wait();
+  return 0;
+}
```
Comment (steffenlarsen) [inline sycl/test/check_device_code/task_sequence_intel_explicit_get.cpp:45]:
```suggestion
}

```

File: sycl/test/check_device_code/task_sequence_intel_no_explicit_get.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                    decltype(properties{pipelined<0>, stall_enable_clusters,
+                                        invocation_capacity<1>,
+                                        response_capacity<1>})>
+          arrayAddTask;
+      arrayAddTask.async(d1, d2, kSize);
+    });
+  });
+  myQueue.wait();
+  return 0;
+}
```
Comment (steffenlarsen) [inline sycl/test/check_device_code/task_sequence_intel_no_explicit_get.cpp:40]:
```suggestion
}

```

File: sycl/test/extensions/properties/properties_task_sequence.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  static_assert(
+      P::has_property<intel::experimental::invocation_capacity_key>());
+  static_assert(P::get_property<intel::experimental::balanced_key>() ==
+                intel::experimental::balanced);
+  static_assert(P::get_property<intel::experimental::response_capacity_key>() ==
+                intel::experimental::response_capacity<TestResponseCapacity>);
+  static_assert(
+      P::get_property<intel::experimental::invocation_capacity_key>() ==
+      intel::experimental::invocation_capacity<TestInvocationCapacity>);
+}
```
Comment (steffenlarsen) [inline sycl/test/extensions/properties/properties_task_sequence.cpp:52]:
```suggestion
}

```

**Review by steffenlarsen — APPROVED** [review]
Overall I think it looks good!

**Review by AlexeySachkov — APPROVED** [review]

**Comment by aejjehint** [comment]
@intel/llvm-gatekeepers PR is ready to merge

**Comment by againull** [comment]
Notified @aejjehint about post-commit failures: https://github.com/intel/llvm/actions/runs/10117762715/job/27983612239

**Pull request merged**